File: 프로그래머스/연습문제/Lv2/두 원 사이의 정수 쌍/두 원 사이의 정수 쌍.py
def solution(r1, r2):
    answer = 8+2*(r2-r1-1)

    r = 1
    while True:
        a = r1**2-r**2

        if a < 0:
            a = 1
            answer += 2
        else:
            sqrt_a = int(a**0.5)
            if sqrt_a**2 != a: sqrt_a += 1
            a = int(sqrt_a)
            
        b = int((r2**2-r**2)**0.5)

        if b == 0: break
        elif a == 0: a += 1

        answer += 4*(b-a+1)
        r += 1

    return answer-2

File: 프로그래머스/연습문제/Lv2/두 원 사이의 정수 쌍/두 원 사이의 정수 쌍.cpp
#include <string>
#include <vector>
#include <cmath>
using namespace std;

long long solution(int r1, int r2) {
    long long answer = 8;   // 두 원 꼭짓점 8개
    
    long long r = 1;
    while(1){
        long long a = (long long)r1*(long long)r1-r*r;
        // a가 제곱수가 아니면 +1
        long long sqrt_a = sqrt(a);
        if(sqrt_a*sqrt_a != a) sqrt_a++;
        
        long long b = sqrt((long long)r2*(long long)r2-r*r);
        
        // b가 0이면 끝
        if(b==0) break;
        // a가 0이면 +1
        else if(sqrt_a==0) sqrt_a++;
        // a가 0보다 작으면 1로 만들고 2개 더하기
        else if(sqrt_a < 0){
            sqrt_a = 1;
            answer += 2;
        }
        
        answer += 4*(b-sqrt_a+1);
        r++;
    }
    
    answer += 2*(r2-r1-1);
    return answer;
}
